server.go

```diff
@@ -4174,13 +4325,303 @@ func (s *server) SubscribeCustomMessages() (*subscribe.Client, error) {
+// temporary-access peer to a restricted-access peer. If no restricted-access
+// slots are available, the peer will be disconnected.
+func (s *server) NewPendingCloseChan(remoteHex string) error {
+	// Ensure that we don't lock banScoreMtx and _then_ lock the server
+	// mutex. This could lead to deadlock.
+	s.banScoreMtx.Lock()
+
+	// Fetch the peer's access status from peerScores.
+	status, found := s.peerScores[remoteHex]
+	if !found {
+		s.banScoreMtx.Unlock()
+
+		// If we didn't find the peer, we'll return an error.
+		return ErrNoPeerScore
+	}
+
+	switch status.state {
+	case peerStatusProtected:
+		s.banScoreMtx.Unlock()
+
+		// If this peer is protected, we don't do anything.
+		return nil
+
+	case peerStatusTemporary:
+		// If this peer is temporary, we need to check if it will revert
+		// to a restricted-access peer.
+		peerCount, found := s.peerCounts[remoteHex]
+		if !found {
+			s.banScoreMtx.Unlock()
+
+			// Error if we did not find any info in peerCounts.
+			return ErrNoPendingPeerInfo
+		}
+		currentNumPending := peerCount.PendingOpenCount - 1
+
+		if currentNumPending == 0 {
+			// Remove the entry from peerCounts.
+			delete(s.peerCounts, remoteHex)
+
+			// If this is the only pending-open channel for this
+			// peer and it's getting removed, attempt to demote
+			// this peer to a restricted peer.
+			if s.numRestricted == s.cfg.NumRestrictedSlots {
+				// There are no available restricted slots, so
+				// we need to disconnect this peer.
+				s.banScoreMtx.Unlock()
+
+				hexBytes, err := hex.DecodeString(remoteHex)
+				if err != nil {
+					return err
+				}
+
+				remotePub, err := btcec.ParsePubKey(hexBytes)
+				if err != nil {
+					return err
+				}
+
+				// The banScoreMtx cannot be held when the
+				// below DisconnectPeer call occurs.
+				return s.DisconnectPeer(remotePub)
+			}
+
+			// Otherwise, there is an available restricted-access
+			// slot, so we can demote this peer.
+			s.peerScores[remoteHex] = peerSlotStatus{
+				state: peerStatusRestricted,
+			}
+
+			// Update numRestricted.
+			s.numRestricted++
+
+			s.banScoreMtx.Unlock()
+
+			return nil
+		}
+
+		// Else, we don't need to demote this peer since it has other
+		// pending-open channels with us.
+		peerCount.PendingOpenCount = currentNumPending
+		s.peerCounts[remoteHex] = peerCount
+
+		s.banScoreMtx.Unlock()
+
+		return nil
+
+	case peerStatusRestricted:
+		s.banScoreMtx.Unlock()
+
+		// This should not be possible. This indicates an error.
+		return fmt.Errorf("invalid peer access state transition")
+
+	default:
+		s.banScoreMtx.Unlock()
+
+		// This should not be possible.
+		return fmt.Errorf("invalid peer access status")
+	}
+}
+
+// NewOpenChan is called when a pending-open channel becomes an open channel
+// (i.e. the funding transaction has confirmed). If the remote peer is a
+// temporary-access peer, it will be promoted to a protected-access peer.
+func (s *server) NewOpenChan(remoteHex string) error {
```

Comment:
Missing unit-test

Reply:
I don't think that we can unit-test this because it is a server method and we can't really abstract this?